File: crates/lisca-workspace/src/paths.rs
use std::path::{Path, PathBuf};

pub const BBOX_DIR: &str = "bbox";
pub const ROI_DIR: &str = "roi";
pub const ALIGN_DIR: &str = "align";
pub const MASK_DIR: &str = "mask";
pub const ANALYSIS_DIR: &str = "analysis";
pub const RESULTS_DIR: &str = "results";
pub const ANNOTATIONS_DIR: &str = "annotations";
pub const ANNOTATION_LABELS_JSON: &str = "labels.json";
pub const TIMESERIES_DIR: &str = "timeseries";
pub const SIDECAR_DIR: &str = "sidecar";
pub const ASSAY_JSON: &str = "assay.json";
pub const INDEX_JSON: &str = "index.json";
pub const POS_PREFIX: &str = "Pos";

/// Reserved filenames under `sidecar/` for a future killing sidecar / workstation.
/// lisca does not write these; they are stable import paths for downstream tools.
pub const SIDECAR_EVENTS_JSON: &str = "events.json";
pub const SIDECAR_TRACES_PARQUET: &str = "traces.parquet";
pub const SIDECAR_TRACKS_CSV: &str = "tracks.csv";
pub const SIDECAR_ENGAGEMENTS_CSV: &str = "engagements.csv";

pub fn pos_name(pos: u32) -> String {
    format!("{POS_PREFIX}{pos}")
}

pub fn bbox_csv_name(pos: u32) -> String {
    format!("{}.csv", pos_name(pos))
}

pub fn roi_tiff_name(roi: u32) -> String {
    format!("Roi{roi}.tif")
}

pub fn align_json_name(pos: u32) -> String {
    format!("{}.json", pos_name(pos))
}

pub fn bbox_dir(workspace: impl AsRef<Path>) -> PathBuf {
    workspace.as_ref().join(BBOX_DIR)
}

pub fn bbox_csv_path(workspace: impl AsRef<Path>, pos: u32) -> PathBuf {
    bbox_dir(workspace).join(bbox_csv_name(pos))
}

pub fn roi_dir(workspace: impl AsRef<Path>) -> PathBuf {
    workspace.as_ref().join(ROI_DIR)
}

pub fn roi_pos_dir(workspace: impl AsRef<Path>, pos: u32) -> PathBuf {
    roi_dir(workspace).join(pos_name(pos))
}

pub fn roi_index_path(workspace: impl AsRef<Path>, pos: u32) -> PathBuf {
    roi_pos_dir(workspace, pos).join(INDEX_JSON)
}

pub fn roi_tiff_path(workspace: impl AsRef<Path>, pos: u32, file_name: &str) -> PathBuf {
    roi_pos_dir(workspace, pos).join(file_name)
}

pub fn align_dir(workspace: impl AsRef<Path>) -> PathBuf {
    workspace.as_ref().join(ALIGN_DIR)
}

pub fn align_json_path(workspace: impl AsRef<Path>, pos: u32) -> PathBuf {
    align_dir(workspace).join(align_json_name(pos))
}

pub fn mask_dir(workspace: impl AsRef<Path>) -> PathBuf {
    workspace.as_ref().join(MASK_DIR)
}

pub fn mask_pos_dir(workspace: impl AsRef<Path>, pos: u32) -> PathBuf {
    mask_dir(workspace).join(pos_name(pos))
}

pub fn analysis_dir(workspace: impl AsRef<Path>) -> PathBuf {
    workspace.as_ref().join(ANALYSIS_DIR)
}

pub fn results_dir(workspace: impl AsRef<Path>) -> PathBuf {
    workspace.as_ref().join(RESULTS_DIR)
}

pub fn assay_json_path(workspace: impl AsRef<Path>) -> PathBuf {
    workspace.as_ref().join(ASSAY_JSON)
}

pub fn annotations_dir(workspace: impl AsRef<Path>) -> PathBuf {
    workspace.as_ref().join(ANNOTATIONS_DIR)
}

pub fn annotation_labels_path(workspace: impl AsRef<Path>) -> PathBuf {
    annotations_dir(workspace).join(ANNOTATION_LABELS_JSON)
}

pub fn timeseries_dir(workspace: impl AsRef<Path>) -> PathBuf {
    workspace.as_ref().join(TIMESERIES_DIR)
}

pub fn timeseries_pos_dir(workspace: impl AsRef<Path>, pos: u32) -> PathBuf {
    timeseries_dir(workspace).join(pos_name(pos))
}

pub fn sidecar_dir(workspace: impl AsRef<Path>) -> PathBuf {
    workspace.as_ref().join(SIDECAR_DIR)
}

pub fn sidecar_path(workspace: impl AsRef<Path>, file_name: &str) -> PathBuf {
    sidecar_dir(workspace).join(file_name)
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::path::Path;

    #[test]
    fn path_builders_use_canonical_folder_names() {
        let root = Path::new("/tmp/ws");
        assert_eq!(bbox_csv_path(root, 3), Path::new("/tmp/ws/bbox/Pos3.csv"));
        assert_eq!(
            roi_index_path(root, 3),
            Path::new("/tmp/ws/roi/Pos3/index.json")
        );
        assert_eq!(
            align_json_path(root, 3),
            Path::new("/tmp/ws/align/Pos3.json")
        );
        assert_eq!(assay_json_path(root), Path::new("/tmp/ws/assay.json"));
        assert_eq!(
            annotation_labels_path(root),
            Path::new("/tmp/ws/annotations/labels.json")
        );
        assert_eq!(
            timeseries_pos_dir(root, 1),
            Path::new("/tmp/ws/timeseries/Pos1")
        );
        assert_eq!(
            sidecar_path(root, SIDECAR_EVENTS_JSON),
            Path::new("/tmp/ws/sidecar/events.json")
        );
    }
}

File: python/src/lisca/core/paths.py
"""Canonical workspace folder names and path builders.

Assay sidecars should import these instead of hard-coding directory names.
Live bbox CSV schema is ``roi, x, y, w, h``; see ``lisca.core.bbox.parse_bbox_csv``.
"""

from __future__ import annotations

from pathlib import Path

BBOX_DIR = "bbox"
ROI_DIR = "roi"
ALIGN_DIR = "align"
MASK_DIR = "mask"
ANALYSIS_DIR = "analysis"
RESULTS_DIR = "results"
ANNOTATIONS_DIR = "annotations"
ANNOTATION_LABELS_JSON = "labels.json"
TIMESERIES_DIR = "timeseries"
SIDECAR_DIR = "sidecar"
ASSAY_JSON = "assay.json"
INDEX_JSON = "index.json"
POS_PREFIX = "Pos"

BBOX_COLUMNS: tuple[str, ...] = ("roi", "x", "y", "w", "h")

# Reserved filenames under `sidecar/` for a future killing sidecar / workstation.
# lisca does not write these; they are stable import paths for downstream tools.
SIDECAR_EVENTS_JSON = "events.json"
SIDECAR_TRACES_PARQUET = "traces.parquet"
SIDECAR_TRACKS_CSV = "tracks.csv"
SIDECAR_ENGAGEMENTS_CSV = "engagements.csv"
OCCUPANCY_PACK_JSON = "occupancy-pack.json"


def pos_name(pos: int) -> str:
    return f"{POS_PREFIX}{pos}"


def bbox_csv_name(pos: int) -> str:
    return f"{pos_name(pos)}.csv"


def roi_tiff_name(roi: int) -> str:
    return f"Roi{roi}.tif"


def align_json_name(pos: int) -> str:
    return f"{pos_name(pos)}.json"


def bbox_dir(workspace: Path) -> Path:
    return Path(workspace) / BBOX_DIR


def bbox_csv_path(workspace: Path, pos: int) -> Path:
    return bbox_dir(workspace) / bbox_csv_name(pos)


def roi_dir(workspace: Path) -> Path:
    return Path(workspace) / ROI_DIR


def roi_pos_dir(workspace: Path, pos: int) -> Path:
    return roi_dir(workspace) / pos_name(pos)


def roi_index_path(workspace: Path, pos: int) -> Path:
    return roi_pos_dir(workspace, pos) / INDEX_JSON


def roi_tiff_path(workspace: Path, pos: int, file_name: str) -> Path:
    return roi_pos_dir(workspace, pos) / file_name


def align_dir(workspace: Path) -> Path:
    return Path(workspace) / ALIGN_DIR


def align_json_path(workspace: Path, pos: int) -> Path:
    return align_dir(workspace) / align_json_name(pos)


def occupancy_pack_path(workspace: Path) -> Path:
    """Workspace-level few-shot occupancy pack (not per Pos)."""
    return align_dir(workspace) / OCCUPANCY_PACK_JSON


def mask_dir(workspace: Path) -> Path:
    return Path(workspace) / MASK_DIR


def mask_pos_dir(workspace: Path, pos: int) -> Path:
    return mask_dir(workspace) / pos_name(pos)


def analysis_dir(workspace: Path) -> Path:
    return Path(workspace) / ANALYSIS_DIR


def results_dir(workspace: Path) -> Path:
    return Path(workspace) / RESULTS_DIR


def assay_json_path(workspace: Path) -> Path:
    return Path(workspace) / ASSAY_JSON


def annotations_dir(workspace: Path) -> Path:
    return Path(workspace) / ANNOTATIONS_DIR


def annotation_labels_path(workspace: Path) -> Path:
    return annotations_dir(workspace) / ANNOTATION_LABELS_JSON


def timeseries_dir(workspace: Path) -> Path:
    return Path(workspace) / TIMESERIES_DIR


def timeseries_pos_dir(workspace: Path, pos: int) -> Path:
    return timeseries_dir(workspace) / pos_name(pos)


def sidecar_dir(workspace: Path) -> Path:
    return Path(workspace) / SIDECAR_DIR


def sidecar_path(workspace: Path, file_name: str) -> Path:
    return sidecar_dir(workspace) / file_name
